src/sfizz/SIMDHelpers.cpp

```diff
@@ -0,0 +1,307 @@
+
+#include "SIMDHelpers.h"
+#include "SIMDConfig.h"
+#include "Debug.h"
+#include "simd/HelpersSSE.h"
+#include "simd/HelpersAVX.h"
+#include "cpuid/cpuinfo.hpp"
+#include <array>
+#include <mutex>
+
+namespace sfz {
+
+template <class T>
+struct SIMDDispatch {
+    constexpr SIMDDispatch() = default;
+    void resetStatus();
+    bool getStatus(SIMDOps op) const;
+    void setStatus(SIMDOps op, bool enable);
+
+    decltype(&writeInterleavedScalar<T>) writeInterleaved = &writeInterleavedScalar<T>;
+    decltype(&readInterleavedScalar<T>) readInterleaved = &readInterleavedScalar<T>;
+    decltype(&gainScalar<T>) gain = &gainScalar<T>;
+    decltype(&gain1Scalar<T>) gain1 = &gain1Scalar<T>;
+    decltype(&divideScalar<T>) divide = &divideScalar<T>;
+    decltype(&multiplyAddScalar<T>) multiplyAdd = &multiplyAddScalar<T>;
+    decltype(&multiplyAdd1Scalar<T>) multiplyAdd1 = &multiplyAdd1Scalar<T>;
+    decltype(&linearRampScalar<T>) linearRamp = &linearRampScalar<T>;
+    decltype(&multiplicativeRampScalar<T>) multiplicativeRamp = &multiplicativeRampScalar<T>;
+    decltype(&addScalar<T>) add = &addScalar<T>;
+    decltype(&add1Scalar<T>) add1 = &add1Scalar<T>;
+    decltype(&subtractScalar<T>) subtract = &subtractScalar<T>;
+    decltype(&subtract1Scalar<T>) subtract1 = &subtract1Scalar<T>;
+    decltype(&copyScalar<T>) copy = &copyScalar<T>;
+    decltype(&cumsumScalar<T>) cumsum = &cumsumScalar<T>;
+    decltype(&diffScalar<T>) diff = &diffScalar<T>;
+    decltype(&meanScalar<T>) mean = &meanScalar<T>;
+    decltype(&meanSquaredScalar<T>) meanSquared = &meanSquaredScalar<T>;
+
+private:
+    std::array<bool, static_cast<unsigned>(SIMDOps::_sentinel)> simdStatus;
+    bool initialized { false };
+    cpuid::cpuinfo info;
+};
+
+
+template <>
+bool SIMDDispatch<float>::getStatus(SIMDOps op) const
+{
+    const unsigned index = static_cast<unsigned>(op);
+    ASSERT(index < simdStatus.size());
+    return simdStatus[index];
+}
+
+template <>
+void SIMDDispatch<float>::setStatus(SIMDOps op, bool enable)
+{
+    const unsigned index = static_cast<unsigned>(op);
+    ASSERT(index < simdStatus.size());
+    simdStatus[index] = enable;
+
+    if (!enable) {
+#define SIMD_OP(opname) case SIMDOps::opname : (opname) = opname ## Scalar<float>; return;
+        switch (op) {
+            default: break;
+            SIMD_OP(writeInterleaved)
+            SIMD_OP(readInterleaved)
+            SIMD_OP(gain)
+            SIMD_OP(gain1)
+            SIMD_OP(divide)
+            SIMD_OP(linearRamp)
+            SIMD_OP(multiplicativeRamp)
+            SIMD_OP(add)
+            SIMD_OP(add1)
+            SIMD_OP(subtract)
+            SIMD_OP(subtract1)
+            SIMD_OP(multiplyAdd)
+            SIMD_OP(multiplyAdd1)
+            SIMD_OP(copy)
+            SIMD_OP(cumsum)
+            SIMD_OP(diff)
+            SIMD_OP(mean)
+            SIMD_OP(meanSquared)
+        }
+#undef SIMD_OP
+    }
+
+#ifdef SFIZZ_HAVE_AVX
```

Comment:
It's wrong. The above checks for -mavx flag to the compiler, which will not be present.
The conditionals should check just the processor type.

Reply:
But then aren't these checks already in cpuinfo? This method will probably never be called in some hot path so the boolean check on `cpuinfo::has_avx()` should be pretty insignificant no?

Reply:
Change to `#if SFIZZ_CPU_FAMILY_X86_64 || SFIZZ_CPU_FAMILY_I386`
There's no need of `SFIZZ_HAVE_AVX` whatsoever, since that ISA is checked at runtime.

Reply:
ok